### PR TITLE
Add Docker setup

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | Sylius/Sylius#9414
| License         | MIT

Replaces #232

### CODE REVIEW

File: app/config/config.yml [hunk 1]
```diff
@@ -31,12 +31,12 @@ framework:
 
 doctrine:
     dbal:
-        driver: '%database_driver%'
         host: '%database_host%'
         port: '%database_port%'
         dbname: '%database_name%'
         user: '%database_user%'
         password: '%database_password%'
+        driver: pdo_mysql
```
Comment (pamil) [inline app/config/config.yml]: Why not keep that parameter?

File: app/config/parameters.yml.dist [hunk 2]
```diff
@@ -10,13 +9,13 @@ parameters:
 
     mailer_transport: '%env(SYLIUS_MAILER_TRANSPORT)%'
     mailer_host: '%env(SYLIUS_MAILER_HOST)%'
+    mailer_port: '%env(SYLIUS_MAILER_PORT)%'
```
Comment (pamil) [inline app/config/parameters.yml.dist]: Could we extract this change to another PR?

File: gulpfile.babel.js [hunk 3]
```diff
@@ -52,6 +52,9 @@ watchShop.description = 'Watch shop asset sources and rebuild on changes.';
 export const build = gulp.parallel(buildAdmin, buildShop);
 build.description = 'Build assets.';
 
+export const watch = gulp.parallel(watchAdmin, watchShop);
+watch.description = 'Watch asset sources and rebuild on changes.';
```
Comment (pamil) [inline gulpfile.babel.js]: Could we extract this change?

File: .travis.yml [hunk 4]
```diff
@@ -85,6 +91,25 @@ script:
     - echo "Testing (Behat - brand new, javascript scenarios; @javascript && ~@todo && ~@cli)" "Sylius"
     - bin/behat --strict --no-interaction -vvv -f progress -p cached --tags="@javascript && ~@todo && ~@cli" || bin/behat --strict --no-interaction -vvv -f progress -p cached --tags="@javascript && ~@todo && ~@cli" --rerun
 
+    - docker-compose pull --ignore-pull-failures
+    - docker-compose build --pull
+    - docker-compose up -d
+    - sleep 60
```
Comment (pamil) [inline .travis.yml]: No idea how to solve that better, just leaving a comment so it's more visible

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,108 @@
+version: '3.4'
```
Comment (pamil) [inline docker-compose.yml]: These two different files for dev and prod would be hard to keep in sync (eg. this version differs). 

@jacquesbh mentioned this on Slack on Friday:
> yes. And it's just a step away to create a `docker-compose.traefik.yml` and explain that `docker-compose -f docker-compose.yml -f docker-compose.traefik.yml` can be use

Can we use an architecture like this for those extra containers (eg. mailhog, nodejs)?

File: docker/php/php-cli.dev.ini [hunk 6]
```diff
@@ -0,0 +1,14 @@
+apc.enable_cli = 1
+date.timezone = ${PHP_DATE_TIMEZONE}
+opcache.enable_cli = 1
```
Comment (pamil) [inline docker/php/php-cli.dev.ini]: Opcache in CLI might cause some issues during development

File: app/config/config.yml (hunk 1 above)
Comment (teohhanhui) [reply app/config/config.yml]: Because it cannot be used in any meaningful way. :smile:

File: app/config/parameters.yml.dist (hunk 2 above)
Comment (teohhanhui) [reply app/config/parameters.yml.dist]: But we need it here...

File: gulpfile.babel.js (hunk 3 above)
Comment (teohhanhui) [reply gulpfile.babel.js]: Sure. But we need it here...

File: docker-compose.yml (hunk 5 above)
Comment (teohhanhui) [reply docker-compose.yml]: It's not just about additional services...

File: docker/php/php-cli.dev.ini (hunk 6 above)
Comment (teohhanhui) [reply docker/php/php-cli.dev.ini]: Such as? :D

**Comment by nietzscheson** [comment]
The Docker configurations for the Sylius project should not be included in the standard version. The reasons that I consider main are:

- There are many out there who do not even use Docker. They do not even use virtual machines.
- The learning curve with Docker is quite complex. So its use should not be mandatory. Not all cloud payment systems allow the use of root to install certain packages on the virtual machine.
- There are other containerization projects such as:
   - Docker Container for Windows (https://github.com/docker/for-win)
   - RKT (https://github.com/rkt/rkt)

If Docker is included in the Standard version, the configurations of the Vagrant configured for Sylius are also included.

In https://github.com/Sylius/Sylius/issues/9414 @stefandoorn adds the following: "can it still be a separate repo? I prefer to have my infra stuff in separate repo, instead of combined with the application itself (SRP). Besides that, not everyone likes or is using Docker, so for these it's not needed to have it in the same repo."

Matter of which I totally agree.

So, all efforts to achieve a good configuration of Docker for Sylius (https://github.com/Sylius/docker) should focus on the repository that already exists and keep it separate.

**Comment by teohhanhui** [comment]
@nietzscheson We've already had that discussion in Sylius/Sylius#9414 and on Slack. I don't see the need to bring it up again here...

**Comment by nietzscheson** [comment]
So is. But this is where the code merges and I see that you want to reiterate what precisely was already discussed on the channel. I just want to make clear my position...

**Comment by teohhanhui** [comment]
> how can one re-use this setup when trying to contribute to Sylius/Sylius? through copy-paste of the docker and docker compose files?

One way is to mount the host directory into, say, `/src/Sylius/Sylius`

> Also, certain services are missing

Could be added in other PR(s)

**Comment by gabiudrescu** [comment]
Second thing I want to bring under your attention: running commands under the root user inside containers is going to mess up permissions on the files on the host.

Especially on development mode, I find it better to use another user inside the container that has UID and GID mapped with the one from the host. On the setup I generally use, I have the following in the docker file:

```
groupadd -g "$APPLICATION_GID" "$APPLICATION_GROUP"
useradd -u "$APPLICATION_UID" --home "/home/$APPLICATION_USER" --create-home --shell /bin/bash --no-user-group "$APPLICATION_GROUP" -g "$APPLICATION_GID"
```

This way, whenever entering the container, I make sure I use the $APPLICATION_USER and everything is OK.

Also, my understanding is that running commands inside docker containers as root may imply a security risk:
 - https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b
 - http://blog.dscpl.com.au/2015/12/don-run-as-root-inside-of-docker.html
 - https://github.com/minio/minio/issues/5742
 - https://www.reddit.com/r/docker/comments/7c51qn/stop_running_application_as_root_user_inside/
 - https://github.com/docker/compose/issues/2380#issuecomment-392530292
 - https://github.com/docker/compose/issues/1532#issuecomment-140160603
 - https://github.com/docker/compose/issues/1377 
 - https://github.com/moby/moby/issues/22415

**Comment by gabiudrescu** [comment]
@teohhanhui 

```
One way is to mount the host directory into, say, /src/Sylius/Sylius
```

can you offer me an example on how one can use this docker setup to test a modification it brought to the Sylius/Sylius vendor file?

**Comment by teohhanhui** [comment]
Switching user in the Dockerfile cannot work, because the Dockerfile has to be portable (and redistributable).

**Comment by jacquesbh** [comment]
### User rights

For the user rights we have this:

Makefile:
```makefile
export USER_UID=$(shell id -u)
example:
	docker-compose ps
```

In our Dockerfiles:
```dockerfile
# Use www-data user
ARG USER_UID=1000
RUN usermod -u $USER_UID www-data
```

By default `www-data` has uid `1000`.
On macOS the user id is `501` usually.

This way, by using a simple variable, we avoid a lot of problems.

We don't have a makefile yet. Why not adding one? It's a good way to improve the developers flow.

### Missing containers

The missing containers are for testing purpose.

This is why I propose to use a `docker-compose.testing.yml`. We don't need to run all the containers all the time, specially Selenium and headless browsers since they are consuming a lot.

**Comment by teohhanhui** [comment]
@jacquesbh Anything that involves setting uid/gid in the Dockerfile is not a solution, because of portability/redistributable

**Comment by jacquesbh** [comment]
This is portable, working on Windows, Linux and macOS. We use it everyday for many projects. So…

**Comment by teohhanhui** [comment]
> Missing containers

Let's keep this first PR small?

**Comment by teohhanhui** [comment]
No, I'm talking about the image being portable. It should be able to be used by any user regardless of their uid/gid on the host, without having to rebuild the image.

**Comment by teohhanhui** [comment]
We should support running with arbitrary uid/gid (via the `--user` flag of `docker run` and its equivalent in docker-compose.yml), but otherwise step down from root if necessary:

https://github.com/docker-library/mysql/issues/397#issuecomment-377288842
https://github.com/docker-library/php/issues/70#issuecomment-396595255

**Comment by teohhanhui** [comment]
I hope we could have this PR merged soon, then others could help to make things better.

File: docker/php/php-cli.dev.ini (hunk 6 above)
Comment (pamil) [reply docker/php/php-cli.dev.ini]: Once we had an issue when phpspec was running unexisting spec which had been removed before, the issue vanished after we disabled opcache.

File: app/config/config.yml (hunk 1 above)
Comment (pamil) [reply app/config/config.yml]: Reverted this change for now.

**Comment by pamil** [comment]
Thank you, Teoh, let's make it the best Docker ever for 1.3 release 🎉

File: docker/php/php-cli.dev.ini (hunk 6 above)
Comment (teohhanhui) [reply docker/php/php-cli.dev.ini]: To be honest, that sounds like a misconfiguration of OpCache, so we should be fine here.
